FEATURE: Add email normalization rules setting

nbianca opened the pull request:
This setting can be used to block use of same email address through
aliases. For example, for some email providers adding dots to the
email address is just an alias. Another alias example is adding
+STRING after the email username.

Normalization rules can be used to remove all dots and +STRING parts
of an email.

SamSaffron commented:
This is so close to what I would like! 

I would prefer an implementation closer to https://github.com/discourse/discourse/commit/6f9177e2ed273ebebb8306299425cbfabbf57101 

That way the original email is always stored correctly and is exactly what the user entered, this simply blocks duplicates thus heavily reducing risk of confusion here.  

When admins look at an email they will still see `sam+100@example.com` and when `sam+111@example.com` tries to register sam will not be allowed.

Additionally using the "extra column" approach it means there is zero risk to existing accounts.

If someone signs up as `sam.11@example.com` we should always email that email. Despite the extra protection in place that blocks `sam.....11+173@example.com` from signing up.

nbianca commented:
That is what I implemented first too, but then we decided that the extra column is not necessary. I reverted the latest changes and I think it is closer to what you suggest.

ZogStriP commented on db/migrate/20211013092406_add_normalized_email_to_user_email.rb:
```diff
@@ -0,0 +1,13 @@
+class AddNormalizedEmailToUserEmail < ActiveRecord::Migration[6.1]
+  def change
+    add_column :user_emails, :normalized_email, :string
+    execute "CREATE INDEX index_user_emails_on_normalized_email ON user_emails (LOWER(normalized_email))"
```
Should this be a `UNIQUE` index?

nbianca replied:
I think that if this index was `UNIQUE` then it would be like the site setting was always true.

ZogStriP replied:
Good point 👍

ZogStriP commented on app/models/user_email.rb:
```diff
@@ -17,6 +18,14 @@ class UserEmail < ActiveRecord::Base
+    self.normalized_email = if self.email.present?
+      username, domain = self.email.split('@', 2)
+      username = username.gsub('.', '').gsub(/\+.*/, '')
+      "#{username}@#{domain}"
```
It's not much use, but if we want to be pedant, we could even `downcase` the domain part of the email

```suggestion
      "#{username}@#{domain.downcase}"
```

nbianca replied:
There is `before_validation :strip_downcase_email` which will do this before this method is called. What I mean is that self.email will be stripped and downcased before this method is called.

ZogStriP replied:
Nice 👍
